HW catalog - supported features vNext

kartben opened the pull request:
https://builds.zephyrproject.io/zephyr/pr/86288/docs/boards/nxp/s32z2xxdc2/doc/index.html#supported-features (note that this is a link for a different, draft, PR that has the same changeset as this PR plus a "hack" to allow the full docgen to run in CI)

Improve the board supported hardware feature catalog generation and `.. zephyr:board-supported-hw` directive:

- Differentiate on-chip vs on-board hardware features
- Added count information when a given IP (compatible) appears multiple
  times.
- Show okay *and* disabled features
- Improve table layout and readability
- Added a legend

<img width="844" alt="image" src="https://github.com/user-attachments/assets/c6e7f40a-377f-4754-9c3a-e059f149e5fa" />

kartben committed:
doc: Enhance board supported hardware directive and catalog generation

Improve the board supported hardware feature catalog generation and
`.. zephyr:board-supported-hw` directive:

- Differentiate on-chip vs on-board hardware features
- Added count information when a given IP (compatible) appears multiple
  times.
- Show okay *and* disabled features
- Improve table layout and readability

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
devicetree: Add filename and line number tracking for nodes & properties

This change enhances the devicetree library by adding support for tracking
the source filename and line number for nodes and properties.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

kartben committed:
doc: _extensions: create dedicated board.css/js files

Don't clutter custom.css with css rules and scripts only used in boards'
doc pages.

Signed-off-by: Benjamin Cabé <benjamin@zephyrproject.org>

pdgendt reviewed (approved):
This looks great!

If there are multiple instances of the same compatible, but one is `okay` and the other `disabled`, how will this show up?

erwango reviewed (changes requested):
This is excellent, but we need to foresee the frequent case of readers not yet used to Zephyr way of describing boards.
Please add a legend to tell readers how to interpret symbols (Green and yellow points, grayed lines)

kartben commented:
> If there are multiple instances of the same compatible, but one is `okay` and the other `disabled`, how will this show up?

Thought I had made "okay" always win in case there's a mix of "okay" and "disabled" but looks like I lost that when cleaning up the code. Will re-introduce it (assuming that's what we want?)

kartben commented:
> > Thought I had made "okay" always win in case there's a mix of "okay" and "disabled" but looks like I lost that when cleaning up the code. Will re-introduce it (assuming that's what we want?)
> 
> That's what I was wondering, maybe we split those in 2 lines in the table?

But that would be two almost identical lines though? Like:

ADC | on-chip | 🟢 nRF ADC node | nordic,nrf-adc
ADC | on-chip | 🟡 nRF ADC node | nordic,nrf-adc

I think knowing that at least one peripheral of a given kind is ready to be used should be enough information for most users?

mathieuchopstm commented:
> > > Thought I had made "okay" always win in case there's a mix of "okay" and "disabled" but looks like I lost that when cleaning up the code. Will re-introduce it (assuming that's what we want?)
> > 
> > 
> > That's what I was wondering, maybe we split those in 2 lines in the table?
> 
> But that would be two almost identical lines though? Like:
> 
> ADC | on-chip | 🟢 nRF ADC node | nordic,nrf-adc ADC | on-chip | 🟡 nRF ADC node | nordic,nrf-adc
> 
> I think knowing that at least one peripheral of a given kind is ready to be used should be enough information for most users?

Maybe the color point could be replaced by a green `active instances` counter, and a yellow `inactive instances` counter?
```
ADC | on-chip | nRF ADC node [x1][x2] | nordic,nrf-adc
```
where `x1` would be green, and `x2` would be yellow. Not sure whether that would be less confusing though.

kartben commented:
@mathieuchopstm @pdgendt thoughts?

<img width="836" alt="image" src="https://github.com/user-attachments/assets/05542c6b-fe9d-4277-9051-5805a42257e2" />

kartben commented:
Added a legend as rightly requested by @erwango ; and implemented a different way to visualize okay/disabled peripherals (show the count for each). See screenshot below.

**Note:** I dropped the temporary commit that was used to test this PR in CI so that the commits left are effectively only the ones that should be merged. If people want to see the stuff live, the same changeset WITH the temporary commit has been submitted in another draft PR and CI docs is e.g. : https://builds.zephyrproject.io/zephyr/pr/86288/docs/boards/bbc/microbit/doc/index.html#supported-features

![image](https://github.com/user-attachments/assets/1d91103c-9373-427d-a751-95292d3439ed)

nashif commented:
i would put the legend below the table and make it a bit smaller, right now it is taking the central stage....

Also, I wonder how this will look with boards with many tagets and variants,  this will end up in many long tables, no?

erwango commented on doc/_extensions/zephyr/domain/__init__.py:
```diff
@@ -797,14 +798,53 @@ def run(self):
+            legend,
+            "2",
+            ["count", "disabled-count"],
+            "Number of instances of the feature that are present but initially disabled.",
```
Nit: 
```suggestion
            "Number of instances of the feature that are available but not enabled.",
```

kartben replied:
ya I don't mind this change - but I think the challenge is that saying a feature is "available" on the board (albeit initially not enabled) might sometimes be promising things that are not true? some peripherals "available [on the soc] but not enabled" may simply be never usable on whatever tiny board with very limited pinout, right? So I think I was trying to convey that.

erwango replied:
For the sake of clarity of this table (and what can be explained in the legend) I think that we should be as close possible to what is actually done in boards description. IE, nodes in .dtsi are available (could potentially be enabled and used) and then made enabled in boards description (and then they are usable "Out of the box").
Consideration about limited resources or promises are another concern.

But this is likely and endless discussion.

erwango reviewed (approved):
Thanks for the legend addition.
It may be possible that greying lines is not required anymore now, but if it remains I would think an explanation would still be required.
Anyway, I won't block anymore at this point.

kartben commented:
> i would put the legend below the table and make it a bit smaller, right now it is taking the central stage....

Ya it's already slightly smaller than the main body but I can try to go even smaller. Agree it's kinda large, I can try to make it smaller.

> Also, I wonder how this will look with boards with many tagets and variants, this will end up in many long tables, no?

Yes, right now multiple targets/variants means multiple times what's often a very similar looking table. See e.g the two targets at https://builds.zephyrproject.io/zephyr/pr/86288/docs/boards/nxp/s32z2xxdc2/doc/index.html#supported-features.

Basically for now I would not recommend folks to use the directive if the board has more than 2 targets -- I am working on an improved UI that will group all targets in tabs or something along those lines

fabiobaltieri reviewed (approved):
Fancy

fabiobaltieri commented:
Well, guess that counts as documentation, let's go.
